File: java/src/main/java/com/iap/trace/Attribution.java
package com.iap.trace;

import java.util.Map;

import com.iap.contracts.Trees;

/**
 * {@code decision_trace.schema.json#/$defs/Attribution}: P&amp;L attribution
 * of one decision in bps of traded notional; signs are contributions
 * (negative = cost) and {@code totalBps} is the sum of the five (1e-9).
 *
 * <p>The pinned decomposition from a signal and a TCA record
 * ({@link #of}): {@code alpha = sign(side) * expected_return * 1e4},
 * {@code spread = -spread_cost_bps}, {@code impact = -impact_bps},
 * {@code fees = -fees_bps}, {@code timing = -timing_cost_bps}. The realized
 * P&amp;L never enters it; the residual is reported beside it, never
 * absorbed.
 */
public record Attribution(double alphaBps, double spreadBps, double impactBps,
        double feesBps, double timingBps, double totalBps) {
    private static final String[] KEYS = {"alpha_bps", "spread_bps",
        "impact_bps", "fees_bps", "timing_bps", "total_bps"};

    public Attribution {
        for (double v : new double[] {alphaBps, spreadBps, impactBps, feesBps,
                timingBps, totalBps}) {
            Trees.finite(v, "Attribution");
        }
        double sum = alphaBps + spreadBps + impactBps + feesBps + timingBps;
        if (Math.abs(sum - totalBps) > 1e-9) {
            throw new IllegalArgumentException(
                    "Attribution: total_bps != sum of components");
        }
    }

    /** The pinned decomposition (see class doc); {@code side} 0 = BID. */
    public static Attribution of(int side, double expectedReturn, TCAResultRec tca) {
        double alpha = (side == 0 ? 1.0 : -1.0) * expectedReturn * 1e4;
        double spread = -tca.spreadCostBps();
        double impact = -tca.impactBps();
        double fees = -tca.feesBps();
        double timing = -tca.timingCostBps();
        return new Attribution(alpha, spread, impact, fees, timing,
                alpha + spread + impact + fees + timing);
    }

    /** Schema-ordered tree. */
    public Map<String, Object> toTree() {
        Map<String, Object> t = Trees.ordered();
        t.put("alpha_bps", alphaBps);
        t.put("spread_bps", spreadBps);
        t.put("impact_bps", impactBps);
        t.put("fees_bps", feesBps);
        t.put("timing_bps", timingBps);
        t.put("total_bps", totalBps);
        return t;
    }

    /** Strict inverse of {@link #toTree}. */
    public static Attribution fromTree(Map<String, Object> t) {
        String p = "Attribution";
        Trees.checkKeys(t, KEYS, p);
        return new Attribution(Trees.num(t, "alpha_bps", p),
                Trees.num(t, "spread_bps", p), Trees.num(t, "impact_bps", p),
                Trees.num(t, "fees_bps", p), Trees.num(t, "timing_bps", p),
                Trees.num(t, "total_bps", p));
    }
}

File: python/src/iap/trace/attribution.py
"""P&L attribution of one decision — the pinned decomposition.

For a parent order with signal ``s`` and TCA result ``t`` (all in bps of
traded notional, contributions signed so that negative = cost):

    alpha_bps  = sign(side) * s.expected_return * 1e4
                 (sign = +1 for BID / buy, -1 for ASK / sell: buying on a
                 positive forecast or selling on a negative one is a
                 positive alpha contribution)
    spread_bps = -t.spread_cost_bps
    impact_bps = -t.impact_bps
    fees_bps   = -t.fees_bps
    timing_bps = -t.timing_cost_bps
    total_bps  = alpha_bps + spread_bps + impact_bps + fees_bps + timing_bps

``total_bps`` is the model's explained P&L, not the realized one.  The
**residual** ``realized_bps - total_bps`` (unexplained by the five terms:
delay and opportunity cost, forecast error, drift after the horizon) is
reported next to the attribution by :func:`attribution_report`, never
folded into a component — a decomposition that always sums to the realized
number would hide exactly the error it is supposed to expose.

The identity ``total == sum`` is enforced by the ``Attribution`` contract
to 1e-9; every term is a finite double.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

from iap.contracts.types import AlphaSignal, Attribution, ParentOrder, Side, TCAResult

__all__ = ["AttributionReport", "attribute", "attribution_report", "residual_bps"]


@dataclass(frozen=True)
class AttributionReport:
    """The attribution, the realized P&L it is compared with, and the
    unexplained residual ``realized_bps - attribution.total_bps``."""

    attribution: Attribution
    realized_bps: float
    residual_bps: float


def _side_sign(side: Side) -> float:
    return 1.0 if side is Side.BID else -1.0


def _check_finite(name: str, value: float) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)) \
            or not math.isfinite(value):
        raise ValueError(f"attribute: {name} must be a finite number, got {value!r}")
    return float(value)


def attribute(parent: ParentOrder, signal: AlphaSignal, tca: TCAResult,
              realized_bps: float) -> Attribution:
    """The pinned decomposition (module doc) for ``parent``.

    ``realized_bps`` is the measured P&L of the order in bps of traded
    notional; it is checked finite here so that an attribution is never
    produced for an order whose outcome was not measured, and the residual
    against it is exposed by :func:`residual_bps` /
    :func:`attribution_report` rather than absorbed into a term.  The
    signal and TCA result must belong to the order's instrument.
    """
    if signal.instrument_id != parent.instrument_id:
        raise ValueError(f"attribute: signal instrument {signal.instrument_id} != "
                         f"order instrument {parent.instrument_id}")
    if tca.parent_order_id != parent.parent_order_id:
        raise ValueError(f"attribute: TCA parent order {tca.parent_order_id} != "
                         f"order {parent.parent_order_id}")
    _check_finite("realized_bps", realized_bps)
    alpha = _side_sign(parent.side) * signal.expected_return * 1e4
    spread = -tca.spread_cost_bps
    impact = -tca.impact_bps
    fees = -tca.fees_bps
    timing = -tca.timing_cost_bps
    return Attribution(alpha_bps=alpha, spread_bps=spread, impact_bps=impact,
                       fees_bps=fees, timing_bps=timing,
                       total_bps=alpha + spread + impact + fees + timing)


def residual_bps(attribution: Attribution, realized_bps: float) -> float:
    """``realized_bps - attribution.total_bps``: the part of the realized
    P&L the five terms do not explain."""
    return _check_finite("realized_bps", realized_bps) - attribution.total_bps


def attribution_report(parent: ParentOrder, signal: AlphaSignal, tca: TCAResult,
                       realized_bps: float) -> AttributionReport:
    """:func:`attribute` plus the residual, reported side by side."""
    attribution = attribute(parent, signal, tca, realized_bps)
    return AttributionReport(attribution=attribution, realized_bps=float(realized_bps),
                             residual_bps=residual_bps(attribution, realized_bps))
